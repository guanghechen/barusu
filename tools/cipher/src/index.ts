export * from './util/buffer'
export * from './util/catalog'
export * from './util/cipher'
export * from './util/error'
export * from './util/event-bus'
export * from './util/io'
export * from './util/logger'
export * from './util/master'
export * from './util/secret'
