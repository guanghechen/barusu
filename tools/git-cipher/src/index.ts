import './env/constant'
export * from './command/decrypt'
export * from './command/encrypt'
export * from './command/init'
export * from './core/command'
export * from './core/decrypt/command'
export * from './core/decrypt/context'
export * from './core/decrypt/processor'
export * from './core/encrypt/command'
export * from './core/encrypt/context'
export * from './core/encrypt/processor'
export * from './core/init/command'
export * from './core/init/context'
export * from './core/init/processor'
export * from './core/option'
export * from './env/constant'
export * from './env/constant'
export * from './env/logger'
export * from './env/util'
export * from './util/buffer'
export * from './util/catalog'
export * from './util/cipher'
export * from './util/events'
export * from './util/io'
export * from './util/secret'
