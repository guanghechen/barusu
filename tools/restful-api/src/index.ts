import './env/constant'

export * from './command/generate'
export * from './command/init'
export * from './command/serve'
export * from './core/command'
export * from './core/generate/command'
export * from './core/generate/context'
export * from './core/generate/processor'
export * from './core/generate/util'
export * from './core/init/command'
export * from './core/init/context'
export * from './core/init/processor'
export * from './core/init/util'
export * from './core/option'
export * from './core/serve/command'
export * from './core/serve/context'
export * from './core/serve/middleware/access-log'
export * from './core/serve/middleware/serve-data-file'
export * from './core/serve/processor'
export * from './env/constant'
export * from './env/logger'
export * from './env/util'
export * from './types/api'
export * from './types/api-item'
export * from './types/api-item-group'
export * from './types/raw/api'
export * from './types/raw/api-item'
export * from './types/raw/api-item-group'
export * from './util/api-parser'
