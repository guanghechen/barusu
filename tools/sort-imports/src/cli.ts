import fs from 'fs-extra'
import globby from 'globby'
import path from 'path'
import { name, version } from '@barusu/tool-sort-imports/package.json'
import { createTopCommand, resolveCommandOptions } from '@barusu/util-cli'
import {
  cover,
  coverBoolean,
  coverNumber,
  isNotEmptyArray,
  isNotEmptyString,
} from '@barusu/util-option'
import { StaticImportStatement } from './index'
import {
  COMMAND_NAME,
  CommandOptions,
  ModuleRankItem,
  defaultCommandOptions,
  logger,
} from './util'


const program = createTopCommand(
  COMMAND_NAME,
  version,
  logger
)

program
  .usage('<workspace> [options]')
  .arguments('<workspace>')
  .option('-P, --pattern <pattern>', 'glob pattern of source file', (val, acc: string[]) => acc.concat(val), [])
  .option('-e, --encoding <encoding>', 'encoding of source file')
  .option('--max-column <maxColumn>', 'maximum column width')
  .option('--indent <indent>', 'indent of source codes')
  .option('--quote <quote>', 'quotation marker surround the module path')
  .option('--semicolon', 'whether to add a semicolon at the end of import/export statement')
  .action(function (_workspaceDir: string, options: CommandOptions) {
    const defaultOptions = resolveCommandOptions(
      logger, name, false,
      defaultCommandOptions, _workspaceDir, options)

    // resolve pattern
    const pattern: string[] = cover<string[]>(
      defaultOptions.pattern, options.pattern, isNotEmptyArray)
    logger.debug('pattern:', pattern)

    // resolve quote
    const quote: string = cover<string>(defaultOptions.quote, options.quote, isNotEmptyString)
    logger.debug('quote:', quote)

    // resolve semicolon
    const semicolon: boolean = coverBoolean(defaultOptions.semicolon, options.semicolon)
    logger.debug('semicolon:', semicolon)

    // resolve indent
    const indent: string = cover<string>(defaultOptions.indent, options.indent, isNotEmptyString)
    logger.debug('indent:', indent.replace(' ', '_'))

    // resolve encoding
    const encoding: string = cover<string>(defaultOptions.encoding, options.encoding, isNotEmptyString)
    logger.debug('encoding:', encoding)

    // resolve maxColumn
    const maxColumn: number = coverNumber(defaultOptions.maxColumn, options.maxColumn)
    logger.debug('maxColumn:', maxColumn)

    const moduleRanks: ModuleRankItem[] = cover<ModuleRankItem[]>(
      defaultOptions.moduleRanks, options.moduleRanks, isNotEmptyArray)
    logger.debug('moduleRanks:', moduleRanks)

    // parse regex in moduleRanks
    if (moduleRanks != null) {
      if (!Array.isArray(moduleRanks)) {
        logger.error('[Bad option] moduleRanks should be array.')
        process.exit(-1)
      }

      for (let i = 0; i < moduleRanks.length; ++i) {
        const item = moduleRanks[i]
        if (
          item == null ||
          (typeof item.regex !== 'string') ||
          (typeof item.rank !== 'number')
        ) {
          logger.error(`[Bad option] missed regex / rank in moduleRanks[${ i }]: ${ JSON.stringify(item) }`)
          process.exit(-1)
        }
        try {
          item.regex = new RegExp(item.regex)
        } catch (error) {
          logger.error('', error)
          process.exit(-1)
        }
      }
    }


    const stat = new StaticImportStatement(
      quote, indent, semicolon, Number.parseInt(maxColumn as unknown as string), moduleRanks)
    globby(pattern, { onlyFiles: true, expandDirectories: false })
      .then(matchedPaths => {
        for (const m of matchedPaths) {
          const absolutePath = path.resolve(m)
          logger.debug('absolutePath:', absolutePath)

          const content: string = fs.readFileSync(absolutePath, encoding).toString()
          const resolvedContent: string = stat.process(content)
          if (content !== resolvedContent) {
            logger.info('processing:', absolutePath)
          }
          fs.writeFileSync(absolutePath, resolvedContent, encoding)
        }
      })
      .catch(err => {
        if (err) {
          throw err
        }
      })
  })
  .parse(process.argv)
