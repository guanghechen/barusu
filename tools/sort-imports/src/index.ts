import a, * as waw from './env/constant'
import {
  ModuleRankItem,
  StaticImportOrExportStatItem,
  compareModulePath,
  createCommentRegex,
  createStaticImportOrExportRegexList,
  defaultModuleRankItems,
  formatImportOrExportStatItem,
  execWithMultipleRegex,
} from './util'
export * as x from'./env/constant'
export * from './env/logger'
export * from './util'


export class StaticImportStatement {
  public readonly quote: string
  public readonly indent: string
  public readonly semicolon: boolean
  public readonly maxColumn: number
  public readonly itemRank: Record<StaticImportOrExportStatItem['type'], number>
  public readonly moduleRanks: ModuleRankItem[]
  public readonly staticImportOrExportRegexList: RegExp[]
  public readonly topCommentRegex: RegExp

  public constructor(
    quote = '\'',
    indent = '  ',
    semicolon = false,
    maxColumn = 100,
    moduleRanks: ModuleRankItem[] = defaultModuleRankItems.concat(),
    staticImportOrExportRegexList: RegExp[] = createStaticImportOrExportRegexList('g'),
    topCommentRegex: RegExp = createCommentRegex(),
  ) {
    for (let i = 0; i < staticImportOrExportRegexList.length; ++i) {
      const regex = staticImportOrExportRegexList[i]
      if (regex.flags.indexOf('g') < 0) {
        // eslint-disable-next-line no-param-reassign
        staticImportOrExportRegexList[i] = new RegExp(regex, regex.flags + 'g')
      }
    }

    this.quote = quote
    this.indent = indent
    this.semicolon = semicolon
    this.maxColumn = Number.isNaN(maxColumn) ? 100 : maxColumn,
    this.staticImportOrExportRegexList = staticImportOrExportRegexList
    this.topCommentRegex = new RegExp('^(' + topCommentRegex.source + '|\\s*)*')
    this.itemRank = {
      'import': 1,
      'export': 2,
    }
    this.moduleRanks = moduleRanks
  }

  public process(content: string): string {
    const self = this
    const items: StaticImportOrExportStatItem[] = []
    const m = self.topCommentRegex.exec(content)
    const firstNonCommentIndex = m == null ? 0 : m[0].length

    let startIndex = firstNonCommentIndex
    while (true) {
      const execResult = execWithMultipleRegex([], content, startIndex)
      if (execResult  == null) break
      const m: RegExpExecArray = execResult.result
      const regex = execResult.regex

      if (!/^[;\s]*$/.test(content.substring(startIndex, m.index))) break
      const { defaultExport, exportN: exportNStr, moduleName, type, remainOfLine } = m.groups!
      const exportN: string[] = exportNStr == null ? [] : exportNStr.split(/\s*,\s*/g)
      const item = {
        defaultExport,
        exportN: exportN.filter(x => /\S/.test(x)).sort(),
        moduleName: moduleName.replace(/([\/\\])\.[\/\\]/g, '$1').replace(/([\/\\])+/g, '$1'),
        type: type as any,
        remainOfLine: remainOfLine as string,
      } as StaticImportOrExportStatItem
      item.fullStatement = self.format(item)
      items.push(item)
      startIndex = regex.lastIndex
    }

    if (startIndex <= firstNonCommentIndex) return content.trimLeft()
    const newStatements: string = items
      .sort(self.compare.bind(self))
      .map(item => item.fullStatement).join('\n')
    return (
      content.slice(0, firstNonCommentIndex)
      + (newStatements + '\n\n\n' + content.slice(startIndex).trimLeft()).trimRight()
      + '\n'
    )
  }

  protected format(item: Omit<StaticImportOrExportStatItem, 'fullStatement'>): string {
    const { quote, indent, semicolon, maxColumn } = this
    return formatImportOrExportStatItem(item, quote, indent, semicolon, maxColumn)
  }

  protected compare(x: StaticImportOrExportStatItem, y: StaticImportOrExportStatItem): number {
    if (x.type !== y.type) {
      return this.itemRank[x.type] - this.itemRank[y.type]
    }
    if (x.moduleName !== y.moduleName) {
      return compareModulePath(x.moduleName, y.moduleName, this.moduleRanks)
    }
    if (x.fullStatement === y.fullStatement) return 0
    return x.fullStatement < y.fullStatement ? -1 : 1
  }
}
