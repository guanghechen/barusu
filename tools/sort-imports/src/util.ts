import { CommandConfigurationOptions } from '@barusu/util-cli'


/**
 * 创建匹配静态 import/export 的正则表达式
 */
export function createStaticImportOrExportRegex(flags?: string): RegExp {
  const typeRegex = /(?<type>import|export)/
  const defaultExportRegex = /(?:\s+(?<defaultExport>[\w*]+(?:\s+as\s+[\w]+)?(?:\s*,\s*[\w*]+\s+as\s+[\w]+)?)(?:\s*,)?)/
  const exportNRegex = /(?:\s+\{\s*(?<exportN>(?:[\w]+(?:\s+as\s+[\w]+)?\s*,\s*)*[\w]+(?:\s+as\s+[\w]+)?(?:\s*,)?)\s*\})/
  const moduleNameRegex = /(?:\s+(?<quote>['"])(?<moduleName>[^'"]+)\k<quote>)(?:\s*;+)?/
  const remainOfLineRegex = /(?<remainOfLine>[^\n]*)/
  const regex = new RegExp(
    typeRegex.source
    + defaultExportRegex.source + '?'
    + exportNRegex.source + '?'
    + /(?:\s+from)/.source + '?'
    + moduleNameRegex.source
    + remainOfLineRegex.source
    , flags)
  return regex
}


/**
 * 创建匹配注释的正则表达式
 * @param flags
 */
export function createCommentRegex(flags?: string): RegExp {
  const inlineCommentRegex = /(?<inlineComment>\/\/[^\n]*)/
  const blockCommentRegex = /(?<blockComment>\/\*[\s\S]*?\*\/)/
  const regex = new RegExp(
    inlineCommentRegex.source
    + '|' + blockCommentRegex.source
    , flags)
  return regex
}


/**
 *
 */
export interface ModuleRankItem {
  /**
   *
   */
  regex: RegExp
  /**
   * Rank of the module matched this.regex
   */
  rank: number
}


export const defaultModuleRankItems: ModuleRankItem[] = [
  { // npm package
    regex: /^[a-zA-Z\d][\w\-.]*/,
    rank: 1.1,
  },
  { // npm scoped package
    regex: /^@[a-zA-Z\d][\w\-.]*\/[a-zA-Z\d][\w\-.]*/,
    rank: 1.2,
  },
  { // paths alias
    regex: /^@\//,
    rank: 2.1,
  },
  { // absolute path
    regex: /^(?:\/|[a-zA-Z]:)/,
    rank: 3.1,
  },
  { // relative path (parent)
    regex: /^[.]{2}[\/\\][^\n]*/,
    rank: 3.2,
  },
  { // relative path
    regex: /^[.][\/\\][^\n]*/,
    rank: 3.3,
  }
]


/**
 * Compare the two module paths and determine which one should be ranked first
 * @param p1          path of module1
 * @param p2          path of module2
 * @param moduleItems module priority infos
 */
export function compareModulePath(
  p1: string,
  p2: string,
  moduleItems: ModuleRankItem[]
): number {
  if (p1 === p2) return 0
  let rankOfP1: number | null = null
  let rankOfP2: number | null = null

  for (const moduleItem of moduleItems) {
    if (rankOfP1 == null && moduleItem.regex.test(p1)) {
      rankOfP1 = moduleItem.rank
    }
    if (rankOfP2 == null && moduleItem.regex.test(p2)) {
      rankOfP2 = moduleItem.rank
    }
  }

  /**
   * If there is only one specified rank in p1 and p2, the one assigned the
   * rank will take precedence.
   *
   * Otherwise, If Both the rank of p1 and p2 are not specified or specified
   * with same rank, then just simply compare their lexicographic order
   */
  if (rankOfP1 != null) {
    if (rankOfP2 == null) return -1
    if (rankOfP1 === rankOfP2) return p1 < p2 ? -1: 1
    return rankOfP1 < rankOfP2 ? -1 : 1
  }
  if (rankOfP2 != null) return 1
  return p1 < p2 ? -1 : 1
}


export interface StaticImportOrExportStatItem {
  type: 'import' | 'export'
  moduleName: string
  fullStatement: string
  exportN: string[]
  remainOfLine: string
  defaultExport?: string
}


/**
 *
 * @param item
 * @param indent
 * @param maxColumn
 */
export function formatImportOrExportStatItem(
  item: Omit<StaticImportOrExportStatItem, 'fullStatement'>,
  quote: string,
  indent: string,
  semicolon: boolean,
  maxColumn: number,
): string {
  const assembleStatement = (multiline: boolean): string => {
    let result: string = item.type
    if (item.defaultExport || item.exportN.length > 0) {
      if (item.defaultExport) {
        result += ' ' + item.defaultExport
        if (item.exportN.length > 0) result += ','
      }
      if (item.exportN.length > 0) {
        if (multiline) {
          result += ' {\n' + item.exportN.map(x => indent + x + ',').join('\n') + '\n}'
        } else {
          result += ` { ${ item.exportN.join(', ') } }`
        }
      }
      result += ' from'
    }
    result += ` ${ quote }${ item.moduleName }${ quote }`
    if (semicolon) result += ';'
    return result + item.remainOfLine
  }

  let result = assembleStatement(false)
  if (result.length >= maxColumn && item.exportN) {
    // 将花括号中的 export 拆成多行
    result = assembleStatement(true)
  }
  return result
}


/**
 *
 */
export interface CommandOptions extends CommandConfigurationOptions {
  /**
   * log level
   * @default undefined
   */
  logLevel?: 'debug' | 'verbose' | 'info' | 'warn' | 'error' | string
  /**
   * glob pattern of source file
   * @default []
   */
  pattern: string[]
  /**
   * encoding of source file
   * @default 'utf-8'
   */
  encoding: string
  /**
   * maximum column width
   * @default 80
   */
  maxColumn: number
  /**
   * indent of import/export statements in source codes
   * @default '  '
   */
  indent: string
  /**
   * quotation marker surround the module path
   * @default '\''
   */
  quote: string
  /**
   * whether to add a semicolon at the end of import/export statement
   * @default false
   */
  semicolon: boolean
  /**
   *
   * @default undefined
   */
  moduleRanks: ModuleRankItem[]
}


export const defaultCommandOptions: CommandOptions = {
  logLevel: undefined,
  pattern: [],
  encoding: 'utf-8',
  maxColumn: 80,
  indent: '  ',
  quote: '\'',
  semicolon: false,
  moduleRanks: [],
}
