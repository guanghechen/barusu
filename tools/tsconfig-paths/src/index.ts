import './env/constant'
export * from './command/main'
export * from './core/command'
export * from './core/main/command'
export * from './core/main/context'
export * from './core/main/processor'
export * from './core/main/processor'
export * from './core/option'
export * from './env/constant'
export * from './env/logger'
export * from './util/module-path'
