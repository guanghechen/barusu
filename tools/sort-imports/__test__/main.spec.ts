import fs from 'fs-extra'
import path from 'path'
import { createLoggerMocker } from '@barusu/util-jest'
import { COMMAND_NAME, execMainCommand, logger } from '../src'


describe('stat', function () {
  const caseRootDirectory = path.resolve(__dirname, 'cases', 'files')
  const configFilepath = path.resolve(__dirname, 'cases', 'config.json')

  const kases = fs.readdirSync(caseRootDirectory)

  for (const kase of kases) {
    const { name: title } = path.parse(kase)
    const projectDir = caseRootDirectory
    const filepath = kase
    const absoluteFilepath = path.resolve(caseRootDirectory, filepath)

    test(title, async function () {
      const workspaceRootDir: string = path.resolve(__dirname, '..')
      const loggerMock = createLoggerMocker({ logger, workspaceRootDir })
      loggerMock.mock()

      const fsWriteFileSyncMock = jest
        .spyOn(fs, 'writeFileSync')
        .mockImplementation(function (
          absolutePath: string | any,
          resolveContent: string | any,
          encoding: string | any
        ) {
          expect(absolutePath).toEqual(absoluteFilepath)
          expect(encoding).toEqual('utf-8')
          expect(resolveContent).toMatchSnapshot(`[${ filepath }] resolved content`)
        })

      const args = [
        '',
        COMMAND_NAME,
        projectDir,
        '--pattern',
        absoluteFilepath,
        '-c',
        configFilepath,
        '--log-level=debug',
      ]

      await execMainCommand(args)

      expect(loggerMock.data()).toMatchSnapshot('log info')
      loggerMock.restore()
      fsWriteFileSyncMock.mockRestore()
    })
  }
})
