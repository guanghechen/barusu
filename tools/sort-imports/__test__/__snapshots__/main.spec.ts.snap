// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stat a: [a.txt] resolved content 1`] = `
"import type { A } from '../core/stat/context'
import type { B } from '../core/stat/context'
import {
  SubCommandExecutor,
  SubCommandMounter,
  SubCommandProcessor,
  createSubCommandExecutor,
  createSubCommandMounter,
} from '@barusu/util-cli'
import {
  SubCommandStatOptions,
  createSubCommandStat,
  createWordStatContextFromOptions,
} from '../core/stat/command'
import { WordStatContext } from '../core/stat/context'
import { WordStatProcessor } from '../core/stat/processor'
import { handleError } from './_util'
"
`;

exports[`stat a: log info 1`] = `
Array [
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mcwd: <workspaceRootDir>[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mworkspace: <workspaceRootDir>/__test__/cases/files[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mconfigPath: [[39m
[38;2;188;162;31m  \\"<workspaceRootDir>/__test__/cases/config.json\\"[39m
[38;2;188;162;31m][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mparasticConfigPath: <workspaceRootDir>/package.json[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mparasticConfigEntry: @barusu/tool-sort-imports[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mencoding: utf-8[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mpattern: [[39m
[38;2;188;162;31m  \\"<workspaceRootDir>/__test__/cases/files/a.txt\\"[39m
[38;2;188;162;31m][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mquote: '[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31msemicolon: false[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mtypeFirst: true[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mindent: __[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mmaxColumn: 80[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mmoduleRanks: [[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^(react|vue|angular)(?:[/\\\\\\\\-][\\\\\\\\w\\\\\\\\-./]*)?$\\",[39m
[38;2;188;162;31m    \\"rank\\": 1.1[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^mocha|chai(?:[/][\\\\\\\\w\\\\\\\\-./]*)?$\\",[39m
[38;2;188;162;31m    \\"rank\\": 1.2[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^[a-zA-Z\\\\\\\\d][\\\\\\\\w\\\\\\\\-.]*\\",[39m
[38;2;188;162;31m    \\"rank\\": 1.3[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^@[a-zA-Z\\\\\\\\d][\\\\\\\\w\\\\\\\\-.]*\\\\\\\\/[a-zA-Z\\\\\\\\d][\\\\\\\\w\\\\\\\\-.]*\\",[39m
[38;2;188;162;31m    \\"rank\\": 1.4[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^@\\\\\\\\/\\",[39m
[38;2;188;162;31m    \\"rank\\": 2.1[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^(?:\\\\\\\\/|[a-zA-Z]:)\\",[39m
[38;2;188;162;31m    \\"rank\\": 3.1[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^[.]{2}[\\\\\\\\/\\\\\\\\\\\\\\\\][^\\\\\\\\n]*\\",[39m
[38;2;188;162;31m    \\"rank\\": 3.2[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^[.][\\\\\\\\/\\\\\\\\\\\\\\\\][^\\\\\\\\n]*\\",[39m
[38;2;188;162;31m    \\"rank\\": 3.3[39m
[38;2;188;162;31m  }[39m
[38;2;188;162;31m][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mabsolutePath: <workspaceRootDir>/__test__/cases/files/a.txt[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;0;255;0minfo [39m [90mbarusu-sort-imports[39m]: [38;2;0;255;0mprocessing: <workspaceRootDir>/__test__/cases/files/a.txt[39m
[38;2;0;255;0m[39m",
  ],
]
`;

exports[`stat b: [b.txt] resolved content 1`] = `
"
import fs from 'fs-extra'
import globby from 'globby'
import path from 'path'
import { logger } from '../../env/logger'
import { createCommentRegex } from '../../util/comment'
import { compareModulePath } from '../../util/module-rank'
import {
  StaticModuleStatementItem,
  createStaticImportOrExportRegexList,
  execWithMultipleRegex,
  extractStaticModuleStatementItem,
  formatStaticModuleStatementItem,
} from '../../util/module-statement'
import './env/constant'
export * from './command/main'
export * from './core/command'
export * from './core/main/command'
export * from './core/main/context'
export * from './core/main/processor'
export * from './core/main/processor'
export * from './core/option'
export * from './env/constant'
export * from './env/logger'
export * from './util/comment'
export * from './util/comment'
export * from './util/module-rank'
export * from './util/module-statement'
"
`;

exports[`stat b: log info 1`] = `
Array [
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mcwd: <workspaceRootDir>[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mworkspace: <workspaceRootDir>/__test__/cases/files[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mconfigPath: [[39m
[38;2;188;162;31m  \\"<workspaceRootDir>/__test__/cases/config.json\\"[39m
[38;2;188;162;31m][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mparasticConfigPath: <workspaceRootDir>/package.json[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mparasticConfigEntry: @barusu/tool-sort-imports[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mencoding: utf-8[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mpattern: [[39m
[38;2;188;162;31m  \\"<workspaceRootDir>/__test__/cases/files/b.txt\\"[39m
[38;2;188;162;31m][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mquote: '[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31msemicolon: false[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mtypeFirst: true[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mindent: __[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mmaxColumn: 80[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mmoduleRanks: [[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^(react|vue|angular)(?:[/\\\\\\\\-][\\\\\\\\w\\\\\\\\-./]*)?$\\",[39m
[38;2;188;162;31m    \\"rank\\": 1.1[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^mocha|chai(?:[/][\\\\\\\\w\\\\\\\\-./]*)?$\\",[39m
[38;2;188;162;31m    \\"rank\\": 1.2[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^[a-zA-Z\\\\\\\\d][\\\\\\\\w\\\\\\\\-.]*\\",[39m
[38;2;188;162;31m    \\"rank\\": 1.3[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^@[a-zA-Z\\\\\\\\d][\\\\\\\\w\\\\\\\\-.]*\\\\\\\\/[a-zA-Z\\\\\\\\d][\\\\\\\\w\\\\\\\\-.]*\\",[39m
[38;2;188;162;31m    \\"rank\\": 1.4[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^@\\\\\\\\/\\",[39m
[38;2;188;162;31m    \\"rank\\": 2.1[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^(?:\\\\\\\\/|[a-zA-Z]:)\\",[39m
[38;2;188;162;31m    \\"rank\\": 3.1[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^[.]{2}[\\\\\\\\/\\\\\\\\\\\\\\\\][^\\\\\\\\n]*\\",[39m
[38;2;188;162;31m    \\"rank\\": 3.2[39m
[38;2;188;162;31m  },[39m
[38;2;188;162;31m  {[39m
[38;2;188;162;31m    \\"regex\\": \\"^[.][\\\\\\\\/\\\\\\\\\\\\\\\\][^\\\\\\\\n]*\\",[39m
[38;2;188;162;31m    \\"rank\\": 3.3[39m
[38;2;188;162;31m  }[39m
[38;2;188;162;31m][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-sort-imports[39m]: [38;2;188;162;31mabsolutePath: <workspaceRootDir>/__test__/cases/files/b.txt[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;0;255;0minfo [39m [90mbarusu-sort-imports[39m]: [38;2;0;255;0mprocessing: <workspaceRootDir>/__test__/cases/files/b.txt[39m
[38;2;0;255;0m[39m",
  ],
]
`;
